File: epi_judge_java_solutions/epi/LongestIncreasingSubarray.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
import epi.test_framework.TestFailure;

import java.util.List;

public class LongestIncreasingSubarray {

  // Represent subarray by starting and ending indices, inclusive.
  private static class Subarray {
    public Integer start;
    public Integer end;

    public Subarray(Integer start, Integer end) {
      this.start = start;
      this.end = end;
    }
  }

  public static Subarray findLongestIncreasingSubarray(List<Integer> A) {

    int maxLength = 1;
    Subarray result = new Subarray(0, 0);
    int i = 0;
    while (i < A.size() - maxLength) {
      // Backward check and skip if A[j - 1] >= A[j].
      boolean isSkippable = false;
      for (int j = i + maxLength; j > i; --j) {
        if (A.get(j - 1) >= A.get(j)) {
          i = j;
          isSkippable = true;
          break;
        }
      }

      // Forward check if it is not skippable.
      if (!isSkippable) {
        i += maxLength;
        while (i < A.size() && A.get(i - 1) < A.get(i)) {
          ++i;
          ++maxLength;
        }
        result = new Subarray(i - maxLength, i - 1);
      }
    }
    return result;
  }

  @EpiTest(testDataFile = "longest_increasing_subarray.tsv")
  public static int findLongestIncreasingSubarrayWrapper(List<Integer> A)
      throws Exception {
    Subarray result = findLongestIncreasingSubarray(A);

    if (result.start < 0 || result.start >= A.size() ||
      result.end < 0 || result.end >= A.size() ||
      result.start > result.end)
      throw new TestFailure("Index out of range");

    for (int i = result.start + 1; i < result.end; ++i) {
      if (A.get(i) <= A.get(i - 1)) {
      throw new TestFailure()
          .withMismatchInfo(
              i, String.format("A[%d] < A[%d]", i - 1, i),
              String.format("%d >= %d", A.get(i - 1), A.get(i)));
      }
    }

    return result.end - result.start + 1;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "LongestIncreasingSubarray.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi_judge_cpp_solutions/longest_increasing_subarray.cc
#include <vector>

#include "test_framework/generic_test.h"

using std::vector;

struct Subarray {
  int start = 0, end = 0;
};

Subarray FindLongestIncreasingSubarray(const vector<int>& A) {
  int max_length = 1;
  Subarray result;
  int i = 0;
  while (i < size(A) - max_length) {
    // Backward check and skip if A[j - 1] >= A[j].
    bool is_skippable = false;
    for (int j = i + max_length; j > i; --j) {
      if (A[j - 1] >= A[j]) {
        i = j;
        is_skippable = true;
        break;
      }
    }

    // Forward check if it is not skippable.
    if (!is_skippable) {
      i += max_length;
      while (i < size(A) && A[i - 1] < A[i]) {
        ++i, ++max_length;
      }
      result = {i - max_length, i - 1};
    }
  }
  return result;
}

int FindLongestIncreasingSubarrayWrapper(const vector<int>& A) {
  Subarray result = FindLongestIncreasingSubarray(A);
  return result.end - result.start + 1;
}

// clang-format off


int main(int argc, char* argv[]) {
  std::vector<std::string> args {argv + 1, argv + argc};
  std::vector<std::string> param_names {"A"};
  return GenericTestMain(args, "longest_increasing_subarray.cc", "longest_increasing_subarray.tsv", &FindLongestIncreasingSubarrayWrapper,
                         DefaultComparator{}, param_names);
}
// clang-format on
